Protobufnative schema type

fjod opened the pull request:
"GetSpecificSchema - надо посмотреть как в джаве, это нужно в авро если сообщение было сериализовано не через текущую схему консьюмера, а какую-то старую - 
не нашел ничего похожего в джаве"

"GenericSchema - подумай сам логически, сможешь ли ты десериализовать сообщение в словарь ключ-значение, если сможешь, то можно реализовать"
в protobuf-net не нашел десериализатора схемы (еле как нашел сериализатор)

Нужно как-то указать пользователям, как генерировать классы сообщений при работе с клиентами на разных языках. 
Пример для proto файла для генерации класса через protoc --java-out

```
syntax = "proto3";
package nativeCheck2;

option java_package = "nativeCheck2";
option java_outer_classname = "Test";

message XXXMessage {
    string foo = 1;
    double bar = 2;
}
```

Класс из dotnet, который будет совместим с java 

```
namespace nativeCheck2
{
    [DataContract]
    public class XXXMessage
    {
        [DataMember(Order = 1)]
        public string foo { get; set; }
        [DataMember(Order = 2)]
        public double bar { get; set; }        
    }
}
```

fjod committed:
protobuf_native works, now it's time to work on GetSpecificSchema and GenericSchema

fjod committed:
interop with java

fjod committed:
removal of test project

Lanayx commented on src/Pulsar.Client/Pulsar.Client.fsproj:
```diff
@@ -37,6 +37,7 @@
     <PackageIconUrl />
   </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="protobuf-net.Reflection" Version="3.0.101" />
```
Can you please use paket instead, just like all other packages.

Lanayx commented on src/Pulsar.Client/Schema/ProtobufNativeSchema.fs:
```diff
@@ -0,0 +1,75 @@
+                let set = FileDescriptorSet()
+                let protoForType = Serializer.GetProto<'T> () //save it to .proto file nearby
+                let currentPath = Path.GetDirectoryName(typeof<'T>.Assembly.Location)       
+                let protoFileName = userClassName + ".proto"
```
Can we avoid writing to the file and write to the memory stream instead?

fjod replied:
I dont think so. This serialiation is based on protobuf-net protogen. Here, in the start of the program it's explicitly needs input files 
https://github.com/protobuf-net/protobuf-net/blob/14da8b12ed5ca62d84aa16066f74349a9ee7c5d8/src/protogen/Program.cs#L22
and 
https://github.com/protobuf-net/protobuf-net/blob/14da8b12ed5ca62d84aa16066f74349a9ee7c5d8/src/protogen/Program.cs#L117

..next, files are added to work on 

https://github.com/protobuf-net/protobuf-net/blob/14da8b12ed5ca62d84aa16066f74349a9ee7c5d8/src/protogen/Program.cs#L215

I guess if we want to use memory instead files it would need rewriting protobuf-net.Reflection for our needs.

Lanayx commented on tests/UnitTests/Internal/SchemaTests.fs:
```diff
@@ -17,18 +17,25 @@ type JsonSchemaTest = { X: string; Y: ResizeArray<int> }
 let tests =
     
-    testList "Schema tests" [
+    ftestList "Schema tests" [
```
Don't forget to remove ftestList in the final version of PR

Lanayx commented:
Also please add integration tests and project for generating classes from schema, this link should help https://dejanstojanovic.net/aspnet/2018/june/generate-c-models-from-protobuf-proto-files-directly-from-visual-studio/
(UPDATE: Probably even better link http://protobuf-net.github.io/protobuf-net/contract_first.html)

fjod committed:
pr fixes

fjod commented:
> Also please add integration tests and project for generating classes from schema, this link should help https://dejanstojanovic.net/aspnet/2018/june/generate-c-models-from-protobuf-proto-files-directly-from-visual-studio/
> (UPDATE: Probably even better link http://protobuf-net.github.io/protobuf-net/contract_first.html)

please explain " generating classes from schema" - I still don't know how to generate something useful from bare byte[] descriptor without Google.protobuf

fjod committed:
integration test (cant run them on my machine due to absence of some crypto certs)

Lanayx commented:
"generating classes from schema" - sorry, I meant from .proto file, usually such files can be shared between projects written in different languages.

fjod committed:
addition of protobuf attributes in message class for F# clients

fjod committed:
creating .proto definition for user message class in-memory instead of file on disk

fjod committed:
added c# codegen from .proto (basically a copy-paste from marc grawell protogen)

Lanayx committed:
Merge branch 'develop' into PROTOBUFNATIVE_schema_type
